### DISCUSSION TITLE
What Happens when a Custom `Authenticator` is "skipped"?

### DISCUSSION THREAD

**Discussion by copiri-six** [symfony/symfony#47384] [Q&A] [answered]
From the [docs](https://symfony.com/doc/current/security/custom_authenticator.html):

```
class ApiKeyAuthenticator extends AbstractAuthenticator
{
    /**
     * Called on every request to decide if this authenticator should be
     * used for the request. Returning `false` will cause this authenticator
     * to be skipped.
     */
    public function supports(Request $request): ?bool
    { ... }
    
    ...
}
```

OK... but what happens when an authenticator is the only one, and it's skipped?  Does the request continue as an unauthenticated request?  Is the request rejected (and if so, with what error code)?  Does something else happen?

To put a finer point on it: if I am creating a custom authenticator that I want to handle every request, should I just hard-code `supports()` to return `true` regardless of what it receives, so that the request is evaluated for authentication?  Is it safe to allow any requests to skip the single authenticator?

Thanks in advance, I can't find a clear answer to this anywhere (docs, discussions, StackOverflow, etc.).

**Comment by MatTheCat** [comment, marked as answer]
> what happens when an authenticator is the only one, and it's skipped?

Same thing when all authenticators are skipped: the user is not authenticated.

> Does the request continue as an unauthenticated request? Is the request rejected (and if so, with what error code)? Does something else happen?

The request will reach the [`AccessListener`](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Security/Http/Firewall/AccessListener.php) which will do nothing if no authentication is required. Otherwise it will throw an `AccessDeniedException` which will be caught by the [`ExceptionListener`](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php). The response will then be created by the configured entry point (see https://symfony.com/doc/current/security/entry_point.html), or an HTTP exception will be thrown if there is none.

> if I am creating a custom authenticator that I want to handle every request, should I just hard-code `supports()` to return `true` regardless of what it receives.

You could do it like this yes. It would then be your full responsibility to throw `AuthenticationException`s accordingly.

>  Is it safe to allow any requests to skip the single authenticator?

Yes, because the request’s handling will depend on your configuration (see https://symfony.com/doc/current/security.html#security-access-control).
